engine/sound/src/sound.cpp

```diff
@@ -866,21 +874,22 @@ namespace dmSound
             float left_scale, right_scale;
             GetPanScale(pan, &left_scale, &right_scale);
 
-            float s = (1.0f - mix) * s1 + mix * s2;
+            float s = (1.0f - mix) * s1 + mix * s2; // resulting destination sample value is a mix two source samples since a kind of franctional indexing is used
             mix_buffer[2 * i] += s * gain * left_scale;
             mix_buffer[2 * i + 1] += s * gain * right_scale;
-
-            prev_index = index;
+            
+            prev_index = index; // keep old index for assertion
             frac += delta;
 
             index += (uint32_t)(frac >> RESAMPLE_FRACTION_BITS);
 
-            frac &= ((1U << RESAMPLE_FRACTION_BITS) - 1U);
+            frac &= ((1U << RESAMPLE_FRACTION_BITS) - 1U); // clear bits higher than 33
```

Comment:
Sorry, @JCash. I don't get it.

Let's go through it. RESAMPLE_FRACTION_BITS = 31. Copying from gdb:

(gdb) print /t (1U << 31)
$3 = 10000000000000000000000000000000
(gdb) print /t ((1U << 31) - 1)
$4 = 1111111111111111111111111111111

That's 31 bits in a row. Then comes 

`frac &= 1111111111111111111111111111111` - typing it as a binary for clarity

It's ANDed with `frac`, which is a 64 bit unsigned. I guess,  `1111111111111111111111111111111` is first implicitly expanded to a 64bit value padding zeros to higher 33 bits (it's unsigned). Then ANDed.

Am i missing something ?

Reply:
First of all, I must have been very tired. Sorry!

The code of course keeps the lower bits (somehow I misread the & operator :/)

And, I was thrown off by the fact that the constant is 32-bit (there's no reason is shouldn't be 64 bit). As such, the 33 didn't make any sense with a 32 bit mask.

Regardless, I wouldn't type the actual value of the constant, since it might be changed in the future, and then the comment will be old.

Reply:
Hey, np. I just want to be sure i understand what we 're doing here, especially in this part of the engine with all this bit fiddling.

Ok. I'll use a more abstract comment. Make sense.